src/drivers/intel/dmic/dmic.c

```diff
@@ -0,0 +1,688 @@
+	 */
+	dmic->startcount++;
+	dmic->gain = q_multsr_sat_32x32(dmic->gain, dmic->gain_coef,
+					Q_SHIFT_GAIN_X_GAIN_COEF);
+
+	/* Gain is stored as Q2.30, while HW register is Q1.19 so shift
+	 * the value right by 11.
+	 */
+	gval = dmic->gain >> 11;
+
+	/* Note that DMIC gain value zero has a special purpose. Value zero
+	 * sets gain bypass mode in HW. Zero value will be applied after ramp
+	 * is complete. It is because exact 1.0 gain is not possible with Q1.19.
+	 */
+	if (gval > DMIC_HW_FIR_GAIN_MAX)
+		gval = 0;
+
+	/* Write gain to registers */
+	for (i = 0; i < DMIC_HW_CONTROLLERS; i++) {
+		if (!dmic->enable[i])
+			continue;
+
+		if (dmic->startcount == DMIC_UNMUTE_CIC)
+			dai_update_bits(dai, base[i] + CIC_CONTROL, CIC_CONTROL_MIC_MUTE_BIT, 0);
+
+		if (dmic->startcount == DMIC_UNMUTE_FIR) {
+			switch (dai->index) {
+			case 0:
+				dai_update_bits(dai, base[i] + FIR_CONTROL_A,
+						FIR_CONTROL_A_MUTE_BIT, 0);
+				break;
+			case 1:
+				dai_update_bits(dai, base[i] + FIR_CONTROL_B,
+						FIR_CONTROL_B_MUTE_BIT, 0);
+				break;
+			}
+		}
+		switch (dai->index) {
+		case 0:
+			val = OUT_GAIN_LEFT_A_GAIN(gval);
+			dai_write(dai, base[i] + OUT_GAIN_LEFT_A, val);
+			dai_write(dai, base[i] + OUT_GAIN_RIGHT_A, val);
+			break;
+		case 1:
+			val = OUT_GAIN_LEFT_B_GAIN(gval);
+			dai_write(dai, base[i] + OUT_GAIN_LEFT_B, val);
+			dai_write(dai, base[i] + OUT_GAIN_RIGHT_B, val);
+			break;
+		}
+	}
+
+
+	spin_unlock(&dai->lock);
+
+	return gval ? SOF_TASK_STATE_RESCHEDULE : SOF_TASK_STATE_COMPLETED;
+}
+
+/* get DMIC hw params */
+static int dmic_get_hw_params(struct dai *dai, struct sof_ipc_stream_params  *params, int dir)
+{
+#if CONFIG_INTEL_DMIC_TPLG_PARAMS
+	return dmic_get_hw_params_computed(dai, params, dir);
+
+#elif CONFIG_INTEL_DMIC_NHLT
+	return dmic_get_hw_params_nhlt(dai, params, dir);
+
+#else
+	return -EINVAL;
+#endif
+}
+
+static int dmic_set_config(struct dai *dai, struct ipc_config_dai *common_config,
+			   void *spec_config)
+{
+	struct dmic_pdata *dmic = dai_get_drvdata(dai);
+	struct sof_ipc_dai_config *config = spec_config;
+	int32_t unmute_ramp_time_ms;
+	int32_t step_db;
+	size_t size;
+	int i;
+	int ret = 0;
+	int di = dai->index;
+
+#if CONFIG_INTEL_DMIC_TPLG_PARAMS
+	int j;
+#endif
+	dai_info(dai, "dmic_set_config()");
+
+	if (config->dmic.driver_ipc_version != DMIC_IPC_VERSION) {
+		dai_err(dai, "dmic_set_config(): wrong ipc version");
+		return -EINVAL;
+	}
+
+	spin_lock(&dai->lock);
+
+	/* Compute unmute ramp gain update coefficient. Use the value from
+	 * topology if it is non-zero, otherwise use default length.
+	 */
+	if (config->dmic.unmute_ramp_time)
+		unmute_ramp_time_ms = config->dmic.unmute_ramp_time;
+	else
+		unmute_ramp_time_ms = LOGRAMP_TIME_MS;
+
+	if (unmute_ramp_time_ms < LOGRAMP_TIME_MIN_MS ||
+	    unmute_ramp_time_ms > LOGRAMP_TIME_MAX_MS) {
+		dai_err(dai, "dmic_set_config(): Illegal ramp time = %d",
+			unmute_ramp_time_ms);
+		ret = -EINVAL;
```

Comment:
@singalsu  there is no unmute_ramp_time_ms for IPC4, could you help to isolate these code ?

Reply:
This a tricky one. The BIOS does not describe how long it takes for the microphone signal to become valid. It depends on microphone component and our HW version. It needs to come from topology in Linux and ChromeOS. It would be possible to move this function to volume component (add the log ramp type) if there's always volume in the pipeline after DMIC DAI. Can the DMIC DAI new() still pass tokens values in addition to the blob?